### PR TITLE
GEODE-9101: fix VisibleNode attribute in MemberMXBean

### PR DESCRIPTION
**Pull request by mkevo**
Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [x] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [x] Is your initial contribution a single, squashed commit?

- [x] Does `gradlew build` run cleanly?

- [x] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, check Concourse for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

File: geode-core/src/main/java/org/apache/geode/distributed/internal/ClusterDistributionManager.java
Comment (kirklund) [inline geode-core/src/main/java/org/apache/geode/distributed/internal/ClusterDistributionManager.java:1885]: I think you're trying to change "nodes" to show the number of "servers" in the cluster but it's a different concept. "nodes" is a statistic in DistributionStats that reflects the number of peer-to-peer processes in the cluster, and we really need this statistic to remain as is.

I recommend adding one or more new attribute(s) to MemberMXBean or DistributedSystemMXBean which reflect the number of "non-locator members" or "servers". There are several related operations on DistributedSystemMXBean such listLocators(), listMembers(), listServers(). All three of these are "nodes".

listMembers() should show you all of the locators and servers (ie, any peer-to-peer process) in the cluster.

listLocators() should should you all of the locators only.

listServers() should show you all of the servers only.

There is a subtle difference between Attributes and Operations in JMX. You probably want an Attribute in this case so that it can be monitored with JMX APIs. But I'll leave further planning to you.

**Review by kirklund — CHANGES_REQUESTED** [review]

File: geode-core/src/main/java/org/apache/geode/distributed/internal/ClusterDistributionManager.java
Comment (mkevo) [reply geode-core/src/main/java/org/apache/geode/distributed/internal/ClusterDistributionManager.java:1885]: I don't want to change "nodes" to show the number of "servers".
How it works by now?
If we have 2 locators and 2 servers the VisibleNodes attribute has the following values:
```
Locator 1: 2 (count Server 1 and Server 2)
Locator 2: 2 (count Server 1 and Server 2)
Server 1: 3 (count Server 2, and twice Server 1)
Server 2: 3 (count  Server 1, and twice Server 2)
```

With my changes it will count every member without itself:
```
Locator 1: 3 (count Locator 2, Server 1, and Server 2)
Locator 2: 3 (count Locator 1, Server 1, and Server 2)
Server 1: 3 (count Locator 1, Locator 2, and Server 2)
Server 2: 3 (count Locator 1, Locator 2, and Server 1)
```
What do you mean by the number of "peer-to-peer" processes in the cluster? 
Is that include both locators and servers?

Comment (jinmeiliao) [reply geode-core/src/main/java/org/apache/geode/distributed/internal/ClusterDistributionManager.java:1885]: From the original code, it looks like it doesn't think a locator should be included in the count, so, shouldn't the correct count should be:

```
Locator 1: 2 (count Server 1 and Server 2) 
Locator 2: 2 (count Server 1 and Server 2)
Server 1: 2 (count Server 1 and Server 2) 
Server 2: 2 (count Server 1 and Server 2)
```
instead?

Comment (mkevo) [reply geode-core/src/main/java/org/apache/geode/distributed/internal/ClusterDistributionManager.java:1885]: In the description of this attribute state the following:
`The current number of nodes in this distributed system visible to this member.`

What are nodes there, both locators and servers or just servers?
Also, should the current server be visible to itself?

If we go with @jinmeiliao proposal we should change the description to be more clear.
`The current number of servers in this distributed system visible to this member.`

Comment (kirklund) [reply geode-core/src/main/java/org/apache/geode/distributed/internal/ClusterDistributionManager.java:1885]: I understand the purpose of the PR better now. However, the PR is still trying to redefine the "nodes" statistic in DistributionStats which is not correct. You should instead add a new attribute to MemberMXBean that returns the count of OTHER members in the cluster.

Here is the definition of the "nodes" statistic:
```
    final String nodesDesc = "The current number of nodes in this distributed system.";
```
So, if you have 2 servers and 2 locators, "nodes" should be 4 within all 4 JVMs.

The earliest version of GemFire had locators which were not members (ie not nodes), but locators now are definitely members (ie nodes). The next type of locators in GemFire were members (ie nodes) but without caches. As of GemFire 7.0, all locators have a cache. As of Geode 1.0, all locators are members with caches. The methods to create that earliest type of locator still exists but are deprecated... so deprecated internal methods is now the only way to create the thin locators that are not peer-to-peer members of the cluster.

Changing the attribute of an MXBean is _different_ from changing the meaning of the statistic. The statistic is not supposed to ignore locators.

**Review by jinmeiliao — APPROVED** [review]

**Comment by mkevo** [comment]
@kirklund , can you re-review it, please?

File: geode-core/src/main/java/org/apache/geode/distributed/internal/ClusterDistributionManager.java
Comment (kirklund) [inline geode-core/src/main/java/org/apache/geode/distributed/internal/ClusterDistributionManager.java:354]: This block of code needs to addNewMember even when the name is "". Why does it need to move it into an else-block?

If you start a Geode process without using `start locator` or `start server`, you can do so without a `name` but it's still a new member.

File: geode-core/src/main/java/org/apache/geode/distributed/internal/ClusterDistributionManager.java
Comment (kirklund) [inline geode-core/src/main/java/org/apache/geode/distributed/internal/ClusterDistributionManager.java:1796]: I think @Bill @bschuchardt or @dschneider-pivotal should review this change instead of me.

File: geode-core/src/main/java/org/apache/geode/distributed/internal/ClusterDistributionManager.java
Comment (kirklund) [reply geode-core/src/main/java/org/apache/geode/distributed/internal/ClusterDistributionManager.java:1883]: I'm marking this as resolved because I think developers who have worked on geode-membership should review it instead of me.

File: geode-core/src/main/java/org/apache/geode/distributed/internal/ClusterDistributionManager.java
Comment (mkevo) [reply geode-core/src/main/java/org/apache/geode/distributed/internal/ClusterDistributionManager.java:354]: In the method, _create_(ClusterDistributionManager.java) it adds itself twice(ClusterDistributionManager constructor and there again). So it increments this attribute for both _addNewMember()_ call.

**Comment by echobravopapa** [comment]
@kirklund I noticed your comments on the JIRA ticket, do you still think this warrants dev list discussion and possibly even an RFC?   @mkevo did you see his comments on the ticket?

**Comment by mkevo** [comment]
Hi @echobravopapa, I think that for adding a new attribute we need to start the discussion on the dev list or write an RFC. But for this case, we just changed the existing attribute to print the correct values. 
Some time ago, when this attribute was added, the locators weren't members/nodes, then it is changed and locators became members/nodes without a cache. From GemFire 7.0, all locators also have a cache and they are members/nodes and we should change this attribute to get the correct value for all visible nodes.

**Comment by echobravopapa** [comment]
@mkevo would you mind pushing an empty commit so the PR pipeline will re-run, thanks.

**Review by kirklund — COMMENTED** [review]
This PR is actually redefining what "node" means in DistributionStats in order to "correct" the value of "nodes" on the MemberMXBean. 

We should not change what "node" means in DistributionStats. The current value of that stat is correct for DistributionStats.

You should add a new attribute to MemberMXBean such as "members" or "servers" or "locators" or all three. But changing the underlying value of "node" is not the correct way to fix the problem in MemberMXBean. The authors of MemberMXBean should not have used "node" from the stats -- they should have used the membership layer to count the number of members, server, and/or locators.

I recommended a dev-list discussion because this change is contentious (I disagree with it) and if you add new attributes to MemberMXBean that is adding new User APIs for which we have always said needs RFC or dev-list discussion.

**Comment by onichols-pivotal** [comment]
looks like this may not be moving forward, can it be closed?

**Comment by mkevo** [comment]
> looks like this may not be moving forward, can it be closed?

I still waiting for an answer on the geode dev mail list. So this cannot be closed.

**Comment by kirklund** [comment]
> > looks like this may not be moving forward, can it be closed?
> 
> I still waiting for an answer on the geode dev mail list. So this cannot be closed.

@mkevo since this is dragging out and you didn't get an answer on dev, I would write up a simple proposal on the wiki (similar to the other proposals) and start a review of that on the dev list.

**Comment by mkevo** [comment]
> > > looks like this may not be moving forward, can it be closed?
> > 
> > 
> > I still waiting for an answer on the geode dev mail list. So this cannot be closed.
> 
> @mkevo since this is dragging out and you didn't get an answer on dev, I would write up a simple proposal on the wiki (similar to the other proposals) and start a review of that on the dev list.

Thanks @kirklund!

**Comment by mkevo** [comment]
> > > looks like this may not be moving forward, can it be closed?
> > 
> > 
> > I still waiting for an answer on the geode dev mail list. So this cannot be closed.
> 
> @mkevo since this is dragging out and you didn't get an answer on dev, I would write up a simple proposal on the wiki (similar to the other proposals) and start a review of that on the dev list.

Hi @kirklund,
What is the status of this activity?

**Comment by mkevo** [comment]
> > > > looks like this may not be moving forward, can it be closed?
> > > 
> > > 
> > > I still waiting for an answer on the geode dev mail list. So this cannot be closed.
> > 
> > 
> > @mkevo since this is dragging out and you didn't get an answer on dev, I would write up a simple proposal on the wiki (similar to the other proposals) and start a review of that on the dev list.
> 
> Hi @kirklund, What is the status of this activity?

Hi @kirklund, any news on this?

**Review by mivanac — APPROVED** [review]

File: geode-core/src/main/java/org/apache/geode/distributed/internal/ClusterDistributionManager.java
Comment (kirklund) [reply geode-core/src/main/java/org/apache/geode/distributed/internal/ClusterDistributionManager.java:1796]: Why is it ok to remove the if-condition from here?

File: geode-core/src/main/java/org/apache/geode/distributed/internal/ClusterDistributionManager.java
Comment (kirklund) [reply geode-core/src/main/java/org/apache/geode/distributed/internal/ClusterDistributionManager.java:354]: We probably need some unit tests for ClusterDistributionManager which would be closer to the actual code than JMX.

I can't sign off on this change for now though because as far as I know the value of `nodes` is correct. I still think you should convince dev-list that the value of `nodes` is incorrect or you should leave the value alone.

File: geode-core/src/main/java/org/apache/geode/distributed/internal/ClusterDistributionManager.java
Comment (mkevo) [reply geode-core/src/main/java/org/apache/geode/distributed/internal/ClusterDistributionManager.java:1796]: The stats should count both locators and servers, as both of them are nodes. With if-condition locators wouldn't count.

File: geode-core/src/main/java/org/apache/geode/distributed/internal/ClusterDistributionManager.java
Comment (mkevo) [reply geode-core/src/main/java/org/apache/geode/distributed/internal/ClusterDistributionManager.java:354]: Hi @kirklund ,
I send a mail to the dev list to discuss what is the right way to fix this. No response.
I wrote an RFC with adding a new attribute. The only one response, to fix this ticket instead of adding a new attribute.

**Review by kirklund — COMMENTED** [review]
@mario Please see GEODE-10358 and the test I wrote in PR #7765 to understand why removing `if (theId.getVmKind() != ClusterDistributionManager.LOCATOR_DM_TYPE) {` will result in redefining the `DistributionStats` `nodes` stat which is incorrect.

I'm going to -1 this PR and the RFC. Thank you, Kirk

**Pull request closed without merging**
